packages/gatsby/src/schema/schema.js

```diff
@@ -0,0 +1,317 @@
+      typeComposer,
+      nodeStore,
+      parentSpan,
+    })
+    await addTypeToRootQuery({ schemaComposer, typeComposer, parentSpan })
+  }
+}
+
+// const addTypeDefs = ({ schemaComposer, typeDefs, parentSpan }) => {
+//   typeDefs.forEach(typeDef => {
+//     schemaComposer.addTypeDefs(typeDef)
+//   })
+// }
+
+const addSetFieldsOnGraphQLNodeTypeFields = ({
+  schemaComposer,
+  nodeStore,
+  parentSpan,
+}) =>
+  Promise.all(
+    Array.from(schemaComposer.values()).map(async tc => {
+      if (
+        tc instanceof schemaComposer.TypeComposer &&
+        hasNodeInterface({ schemaComposer, typeComposer: tc })
+      ) {
+        const typeName = tc.getTypeName()
+        const [fields] = await apiRunner(`setFieldsOnGraphQLNodeType`, {
+          type: {
+            name: typeName,
+            nodes: nodeStore.getNodesByType(typeName),
+          },
+          traceId: `initial-setFieldsOnGraphQLNodeType`,
+          parentSpan: parentSpan,
+        })
+        if (fields) {
+          // NOTE: `setFieldsOnGraphQLNodeType` only allows setting
+          // nested fields with a path as property name, i.e.
+          // `{ 'frontmatter.published': 'Boolean' }`, but not in the form
+          // `{ frontmatter: { published: 'Boolean' }}`
+          tc.addNestedFields(fields)
+        }
+      }
+    })
+  )
+
+const addThirdPartySchemas = ({
+  schemaComposer,
+  thirdPartySchemas,
+  parentSpan,
+}) => {
+  thirdPartySchemas.forEach(schema => {
+    const QueryTC = schemaComposer.TypeComposer.createTemp(
+      schema.getQueryType()
+    )
+    const fields = QueryTC.getFields()
+    // TODO: Wrap field resolvers to include projected fields in the
+    // selection set.
+    schemaComposer.Query.addFields(fields)
+
+    // Explicitly add the third-party schema's types, so they can be targeted
+    // in `addResolvers` API.
+    const rootTypeName = Object.values(fields)[0].type.name
+    const types = schema.getTypeMap()
+    Object.entries(types).forEach(
+      ([typeName, type]) =>
+        typeName.startsWith(rootTypeName) && schemaComposer.add(type)
+    )
+  })
+}
+
+// const addCustomResolveFunctions = ({
+//   schemaComposer,
+//   resolvers,
+//   parentSpan,
+// }) => {
+//   Object.entries(resolvers).forEach(([typeName, fields]) => {
+//     if (schemaComposer.has(typeName)) {
+//       const tc = schemaComposer.getTC(typeName)
+//       Object.entries(fields).forEach(([fieldName, resolve]) => {
+//         if (tc.hasField(fieldName)) {
+//           const resolver = tc.getField(fieldName).resolve
+//           tc.extendField(fieldName, {
+//             resolve,
+//           })
+//         } else {
+//           const fieldConfig = resolve
+//           tc.addFields({ [fieldName]: fieldConfig })
+//         }
+//       })
+//     }
+//   })
+// }
+
+const addResolvers = ({ schemaComposer, typeComposer }) => {
+  const typeName = typeComposer.getTypeName()
+  const inputTypeComposer = toInputObjectType(
```

Comment:
Note that currently we exclude `parent` and `children` fields from the InputObjectType

Reply:
You mean in master?

Reply:
Yes in master. But we now can handle the circular structure of NodeInput